src/store/upload-store-level.ts

```diff
@@ -0,0 +1,196 @@
+import type { ImportResult } from 'ipfs-unixfs-importer';
+import type { Readable } from 'readable-stream';
+import type { UnixFSEntry } from 'ipfs-unixfs-exporter';
+import type { UploadCompleteResult, UploadPartResult, UploadStore, UploadStoreOptions } from './upload-store.js';
+
+import { abortOr } from '../utils/abort.js';
+import { BlockstoreLevel } from './blockstore-level.js';
+import { Cid } from '../utils/cid.js';
+import { createLevelDatabase } from './level-wrapper.js';
+import { DataStream } from '../utils/data-stream.js';
+import { Encoder } from '../utils/encoder.js';
+import { exporter } from 'ipfs-unixfs-exporter';
+import { importer } from 'ipfs-unixfs-importer';
+import { sum } from '../utils/array.js';
+
+const DATA_PARTITION = 'data';
+const ROOT_PARTITION = 'root';
+
+/**
+ * A simple implementation of {@link UploadStore} that works in both the browser and server-side.
+ * Leverages LevelDB under the hood.
+ *
+ * It has the following structure (`+` represents a sublevel and `->` represents a key->value pair):
+ *   '<tenant>' + <recordId> + 'data' + <index> -> <data>
+ *   '<tenant>' + <recordId> + 'root' + <index> -> <dataCid>
+ *
+ * This allows for the <data> to be stored without having to provide it's `<dataCid>` upfront, which
+ * is especially necessary when getting _all_ the data for a <recordId> for a <tenant>, as the root
+ * <dataCid> can be retrieved for each <index> internally (i.e. not needing the client to provide
+ * each <dataCid> for each part).
+ */
+export class UploadStoreLevel implements UploadStore {
+  config: UploadStoreLevelConfig;
+
+  blockstore: BlockstoreLevel;
+
+  constructor(config: UploadStoreLevelConfig = { }) {
+    this.config = {
+      blockstoreLocation: 'UPLOADSTORE',
+      createLevelDatabase,
+      ...config
+    };
+
+    this.blockstore = new BlockstoreLevel({
+      location            : this.config.blockstoreLocation,
+      createLevelDatabase : this.config.createLevelDatabase,
+    });
+  }
+
+  async open(): Promise<void> {
+    await this.blockstore.open();
+  }
+
+  async close(): Promise<void> {
+    await this.blockstore.close();
+  }
+
+  async start(_tenant: string, _recordId: string, _dataFormat: string, options?: UploadStoreOptions): Promise<boolean> {
```

Comment:
What is the use case for having `start()` at all if it's no-op? I see that a developer may want to supply their own `upload-store` with a custom `start()`, but what would they use it for? If there are some ideas you have in mind, I'm down to keep this, but otherwise I'm partial to omitting it until that need pops up.